# Lab 09 - Data Mining - Ethan Pham

- For every output that adds a new knowledge to our data process, you need to have an "Observation".

## Introduction

1. Context: There are three datasets in this lab, movie.csv, ratings.csv, and user.csv, this lab plans to use all of it. The movie dataset details a large list of movies that are categorized into different genres. The ratings dataset details a long list of ratings assigned to the movie id's. And the user dataset details a long list of users who wrote those reviews.
2. Objectives: The main objective of this lab is to get into the main operations of data mining. Using functions like groupby() and merge, the datasets can be combined together to improve data mining. 
3. Data Description: 
- Movies Dataset:
    - movie id: This is the ID of a movie in the dataset. 
    - movie title: This is the title of a movie in the dataset.
    - release date: This is the release date of a movie in the dataset.
    - Various Genres: There are various genres in this dataset that correspond to a movie. The value 1 in a genre means that the movie has this genre, the value 0 means the movie does not have this genre.
    
- Ratings Dataset:
    - user id: This is the ID of a user who reviews a movie in the dataset.
    - movie id: This is the ID of a movie in the dataset.
    - rating: This is the rating of a movie.
    - timestamp: This is the time when the rating was submitted.
- User Dataset:
    - user id: This is the ID of a user who reviews a movie in the dataset.
    - age: This is the age of a user who reviews a movie in the dataset.
    - gender: This is the gender of a user who reviews a movie in the dataset.
    - occupation: This is the occupation(Job) of a user who reviews a movie in the dataset.
    - zip code: This is the zip code of a user who reviews a movie in the dataset.

## Importing Libararies

In [1]:
# general libraries
import numpy as np   # for numeric values
import pandas as pd  # for text values
import math          # for any math operations
import os

# visualization libraries
import matplotlib.pyplot as plt
import seaborn as sns

# supressing any warnings
import warnings
warnings.filterwarnings('ignore')

## 3. Loading Dataset

In [2]:
df = pd.read_csv('movie.csv')
df1 = pd.read_csv('ratings.csv')
df2 = pd.read_csv('user.csv')
movie = df.copy()
ratings = df1.copy()
user = df2.copy()

## 4. EDA

### Basic Data Exploration
1. head, tail.
2. shape, ndim, size, dtypes, columns (data attributes)
3. missing values
4. dataset information
5. statistical summary
6. object summary

In [3]:
movie.head()

movie id  movie title release date  Action  Adventure  Animation  \
0         1   Toy Story      1-Jan-95       0          0          1   
1         2   GoldenEye      1-Jan-95       1          1          0   
2         3  Four Rooms      1-Jan-95       0          0          0   
3         4  Get Shorty      1-Jan-95       1          0          0   
4         5     Copycat      1-Jan-95       0          0          0   

   Childrens  Comedy  Crime  Documentary  ...  Fantasy  Film-Noir  Horror  \
0          1       1      0            0  ...        0          0       0   
1          0       0      0            0  ...        0          0       0   
2          0       0      0            0  ...        0          0       0   
3          0       1      0            0  ...        0          0       0   
4          0       0      1            0  ...        0          0       0   

   Musical  Mystery  Romance  Sci-Fi  Thriller  War  Western  
0        0        0        0       0         0    0        0  
1        0        0        0       0         1    0        0  
2        0        0        0       0         1    0        0  
3        0        0        0       0         0    0        0  
4        0        0        0       0         1    0        0  

[5 rows x 21 columns]

In [4]:
ratings.head()

user id  movie id  rating  timestamp
0      196       242       3  881250949
1      186       302       3  891717742
2       22       377       1  878887116
3      244        51       2  880606923
4      166       346       1  886397596

In [5]:
user.head()

user id  age gender  occupation zip code
0        1   24      M  technician    85711
1        2   53      F       other    94043
2        3   23      M      writer    32067
3        4   24      M  technician    43537
4        5   33      F       other    15213

In [6]:
movie.shape

(1680, 21)

In [7]:
ratings.shape

(100000, 4)

In [8]:
user.shape

(943, 5)

**Observation:** (2)
1. There are 1680 rows in the movie dataset, 100000 rows in the ratings dataset, and 943 rows in the user dataset.
2. There are 21 columns in the movie dataset, 4 columns in the ratings dataset, and 5 columns in the user dataset.

In [9]:
movie.ndim

2

In [10]:
ratings.ndim

2

In [11]:
user.ndim

2

In [12]:
movie.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1680 entries, 0 to 1679
Data columns (total 21 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   movie id      1680 non-null   int64 
 1   movie title   1680 non-null   object
 2   release date  1680 non-null   object
 3   Action        1680 non-null   int64 
 4   Adventure     1680 non-null   int64 
 5   Animation     1680 non-null   int64 
 6   Childrens     1680 non-null   int64 
 7   Comedy        1680 non-null   int64 
 8   Crime         1680 non-null   int64 
 9   Documentary   1680 non-null   int64 
 10  Drama         1680 non-null   int64 
 11  Fantasy       1680 non-null   int64 
 12  Film-Noir     1680 non-null   int64 
 13  Horror        1680 non-null   int64 
 14  Musical       1680 non-null   int64 
 15  Mystery       1680 non-null   int64 
 16  Romance       1680 non-null   int64 
 17  Sci-Fi        1680 non-null   int64 
 18  Thriller      1680 non-null   int64 
 19  War   

In [13]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype
---  ------     --------------   -----
 0   user id    100000 non-null  int64
 1   movie id   100000 non-null  int64
 2   rating     100000 non-null  int64
 3   timestamp  100000 non-null  int64
dtypes: int64(4)
memory usage: 3.1 MB


In [14]:
user.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 943 entries, 0 to 942
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   user id     943 non-null    int64 
 1   age         943 non-null    int64 
 2   gender      943 non-null    object
 3   occupation  943 non-null    object
 4   zip code    943 non-null    object
dtypes: int64(2), object(3)
memory usage: 37.0+ KB


In [15]:
movie.describe().T

count        mean         std  min     25%    50%      75%  \
movie id     1680.0  841.525595  485.609591  1.0  421.75  841.5  1261.25   
Action       1680.0    0.149405    0.356593  0.0    0.00    0.0     0.00   
Adventure    1680.0    0.080357    0.271926  0.0    0.00    0.0     0.00   
Animation    1680.0    0.025000    0.156171  0.0    0.00    0.0     0.00   
Childrens    1680.0    0.072619    0.259587  0.0    0.00    0.0     0.00   
Comedy       1680.0    0.300595    0.458653  0.0    0.00    0.0     1.00   
Crime        1680.0    0.064881    0.246389  0.0    0.00    0.0     0.00   
Documentary  1680.0    0.029762    0.169980  0.0    0.00    0.0     0.00   
Drama        1680.0    0.431548    0.495440  0.0    0.00    0.0     1.00   
Fantasy      1680.0    0.013095    0.113717  0.0    0.00    0.0     0.00   
Film-Noir    1680.0    0.014286    0.118701  0.0    0.00    0.0     0.00   
Horror       1680.0    0.054762    0.227583  0.0    0.00    0.0     0.00   
Musical      1680.0    0.033333    0.179559  0.0    0.00    0.0     0.00   
Mystery      1680.0    0.036310    0.187115  0.0    0.00    0.0     0.00   
Romance      1680.0    0.147024    0.354235  0.0    0.00    0.0     0.00   
Sci-Fi       1680.0    0.060119    0.237778  0.0    0.00    0.0     0.00   
Thriller     1680.0    0.149405    0.356593  0.0    0.00    0.0     0.00   
War          1680.0    0.042262    0.201246  0.0    0.00    0.0     0.00   
Western      1680.0    0.016071    0.125788  0.0    0.00    0.0     0.00   

                max  
movie id     1682.0  
Action          1.0  
Adventure       1.0  
Animation       1.0  
Childrens       1.0  
Comedy          1.0  
Crime           1.0  
Documentary     1.0  
Drama           1.0  
Fantasy         1.0  
Film-Noir       1.0  
Horror          1.0  
Musical         1.0  
Mystery         1.0  
Romance         1.0  
Sci-Fi          1.0  
Thriller        1.0  
War             1.0  
Western         1.0

In [16]:
ratings.describe()

user id       movie id         rating     timestamp
count  100000.00000  100000.000000  100000.000000  1.000000e+05
mean      462.48475     425.530130       3.529860  8.835289e+08
std       266.61442     330.798356       1.125674  5.343856e+06
min         1.00000       1.000000       1.000000  8.747247e+08
25%       254.00000     175.000000       3.000000  8.794487e+08
50%       447.00000     322.000000       4.000000  8.828269e+08
75%       682.00000     631.000000       4.000000  8.882600e+08
max       943.00000    1682.000000       5.000000  8.932866e+08

In [17]:
user.describe()

user id         age
count  943.000000  943.000000
mean   472.000000   34.051962
std    272.364951   12.192740
min      1.000000    7.000000
25%    236.500000   25.000000
50%    472.000000   31.000000
75%    707.500000   43.000000
max    943.000000   73.000000

## Data Mining Questions

### 1. How many movies are there per each genre?

In [18]:
movie[['Action', 'Adventure', 'Animation', 'Childrens', 'Comedy', 'Crime', 'Documentary', 'Drama', 'Fantasy',
       'Film-Noir', 'Horror', 'Musical', 'Mystery', 'Romance', 'Sci-Fi', 'Thriller', 'War', 'Western']].sum()

Action         251
Adventure      135
Animation       42
Childrens      122
Comedy         505
Crime          109
Documentary     50
Drama          725
Fantasy         22
Film-Noir       24
Horror          92
Musical         56
Mystery         61
Romance        247
Sci-Fi         101
Thriller       251
War             71
Western         27
dtype: int64

In [19]:
genre = movie.loc[:, 'Action':'Western'].sum()
genre

Action         251
Adventure      135
Animation       42
Childrens      122
Comedy         505
Crime          109
Documentary     50
Drama          725
Fantasy         22
Film-Noir       24
Horror          92
Musical         56
Mystery         61
Romance        247
Sci-Fi         101
Thriller       251
War             71
Western         27
dtype: int64

### 2. What are the top 10 genres?

In [20]:
# sort_values() method can sort the values for us; however,
# 1. Without any parameter, the values will be listed ascendingly (A-Z)
# With the (ascending=False), the values will descend (Z-A)

movie.loc[:, 'Action':'Western'].sum().sort_values(ascending=False).head()

Drama       725
Comedy      505
Action      251
Thriller    251
Romance     247
dtype: int64

### 3. What are the movies with more than one genre?
1. Creating a new datast based in the movie id and movie title columns
2. Adding a new column to the dataset num of genres
3. Getting the movies with more than one genre

In [21]:
# 1.
genre_movie = movie[['movie id', 'movie title']]
genre_movie

movie id                          movie title
0            1                           Toy Story 
1            2                           GoldenEye 
2            3                          Four Rooms 
3            4                          Get Shorty 
4            5                             Copycat 
...        ...                                  ...
1675      1678                          Mat' i syn 
1676      1679                           B. Monkey 
1677      1680                       Sliding Doors 
1678      1681                        You So Crazy 
1679      1682  Scream of Stone (Schrei aus Stein) 

[1680 rows x 2 columns]

In [22]:
# 2.
genre_movie['num of genres'] = movie.loc[:, 'Action':'Western'].sum(axis=1)

In [23]:
genre_movie

movie id                          movie title  num of genres
0            1                           Toy Story               3
1            2                           GoldenEye               3
2            3                          Four Rooms               1
3            4                          Get Shorty               3
4            5                             Copycat               3
...        ...                                  ...            ...
1675      1678                          Mat' i syn               1
1676      1679                           B. Monkey               2
1677      1680                       Sliding Doors               2
1678      1681                        You So Crazy               1
1679      1682  Scream of Stone (Schrei aus Stein)               1

[1680 rows x 3 columns]

In [24]:
# 3.
genre_movie[genre_movie['num of genres']>3]

movie id              movie title  num of genres
16          17     From Dusk Till Dawn               5
20          21  Muppet Treasure Island               5
28          29          Batman Forever               4
49          50               Star Wars               5
50          51     Legends of the Fall               4
...        ...                      ...            ...
1108      1110               Tank Girl               4
1136      1138                Best Men               4
1217      1219          Goofy Movie, A               4
1291      1293                Star Kid               4
1612      1615      Warriors of Virtue               4

[65 rows x 3 columns]

### 4. What are the first 25 top-rated movies with 100 or more reviews?
- We want the movies that have highest average of rating.
- Also, those movies with the highest number of reviews.
- We will use the movie and the ratings datasets to find the answer.
- We use the Pandas merge() and groupby():
    - merge() to merge the datasets.
    - groupby() to group the variables.


Steps:
1. merge the datasets and name it merged_df.
2. groupby() the "movie title", and "rating", and name it rating_count.
3. get the average of all the ratings for each movie, and name it avg_rating.
4. filter the movies with more than 100 reviews, and name it movie_100_reviews.
5. use the isin() to check inside both avg_rating, and movie_100_reviews variables to get the movies that answer the question.




1. Merging two datasets
2. Grouping the datasets

In [25]:
movie.head()

movie id  movie title release date  Action  Adventure  Animation  \
0         1   Toy Story      1-Jan-95       0          0          1   
1         2   GoldenEye      1-Jan-95       1          1          0   
2         3  Four Rooms      1-Jan-95       0          0          0   
3         4  Get Shorty      1-Jan-95       1          0          0   
4         5     Copycat      1-Jan-95       0          0          0   

   Childrens  Comedy  Crime  Documentary  ...  Fantasy  Film-Noir  Horror  \
0          1       1      0            0  ...        0          0       0   
1          0       0      0            0  ...        0          0       0   
2          0       0      0            0  ...        0          0       0   
3          0       1      0            0  ...        0          0       0   
4          0       0      1            0  ...        0          0       0   

   Musical  Mystery  Romance  Sci-Fi  Thriller  War  Western  
0        0        0        0       0         0    0        0  
1        0        0        0       0         1    0        0  
2        0        0        0       0         1    0        0  
3        0        0        0       0         0    0        0  
4        0        0        0       0         1    0        0  

[5 rows x 21 columns]

In [26]:
ratings.head()

user id  movie id  rating  timestamp
0      196       242       3  881250949
1      186       302       3  891717742
2       22       377       1  878887116
3      244        51       2  880606923
4      166       346       1  886397596

In [27]:
ratings

user id  movie id  rating  timestamp
0          196       242       3  881250949
1          186       302       3  891717742
2           22       377       1  878887116
3          244        51       2  880606923
4          166       346       1  886397596
...        ...       ...     ...        ...
99995      880       476       3  880175444
99996      716       204       5  879795543
99997      276      1090       1  874795795
99998       13       225       2  882399156
99999       12       203       3  879959583

[100000 rows x 4 columns]

In [28]:
ratings.shape

(100000, 4)

In [29]:
# 1. merging the two datasets
merged_df = movie.merge(ratings, on='movie id')

In [30]:
merged_df.head(20)

movie id movie title release date  Action  Adventure  Animation  \
0          1  Toy Story      1-Jan-95       0          0          1   
1          1  Toy Story      1-Jan-95       0          0          1   
2          1  Toy Story      1-Jan-95       0          0          1   
3          1  Toy Story      1-Jan-95       0          0          1   
4          1  Toy Story      1-Jan-95       0          0          1   
5          1  Toy Story      1-Jan-95       0          0          1   
6          1  Toy Story      1-Jan-95       0          0          1   
7          1  Toy Story      1-Jan-95       0          0          1   
8          1  Toy Story      1-Jan-95       0          0          1   
9          1  Toy Story      1-Jan-95       0          0          1   
10         1  Toy Story      1-Jan-95       0          0          1   
11         1  Toy Story      1-Jan-95       0          0          1   
12         1  Toy Story      1-Jan-95       0          0          1   
13         1  Toy Story      1-Jan-95       0          0          1   
14         1  Toy Story      1-Jan-95       0          0          1   
15         1  Toy Story      1-Jan-95       0          0          1   
16         1  Toy Story      1-Jan-95       0          0          1   
17         1  Toy Story      1-Jan-95       0          0          1   
18         1  Toy Story      1-Jan-95       0          0          1   
19         1  Toy Story      1-Jan-95       0          0          1   

    Childrens  Comedy  Crime  Documentary  ...  Musical  Mystery  Romance  \
0           1       1      0            0  ...        0        0        0   
1           1       1      0            0  ...        0        0        0   
2           1       1      0            0  ...        0        0        0   
3           1       1      0            0  ...        0        0        0   
4           1       1      0            0  ...        0        0        0   
5           1       1      0            0  ...        0        0        0   
6           1       1      0            0  ...        0        0        0   
7           1       1      0            0  ...        0        0        0   
8           1       1      0            0  ...        0        0        0   
9           1       1      0            0  ...        0        0        0   
10          1       1      0            0  ...        0        0        0   
11          1       1      0            0  ...        0        0        0   
12          1       1      0            0  ...        0        0        0   
13          1       1      0            0  ...        0        0        0   
14          1       1      0            0  ...        0        0        0   
15          1       1      0            0  ...        0        0        0   
16          1       1      0            0  ...        0        0        0   
17          1       1      0            0  ...        0        0        0   
18          1       1      0            0  ...        0        0        0   
19          1       1      0            0  ...        0        0        0   

    Sci-Fi  Thriller  War  Western  user id  rating  timestamp  
0        0         0    0        0      308       4  887736532  
1        0         0    0        0      287       5  875334088  
2        0         0    0        0      148       4  877019411  
3        0         0    0        0      280       4  891700426  
4        0         0    0        0       66       3  883601324  
5        0         0    0        0        5       4  875635748  
6        0         0    0        0      109       4  880563619  
7        0         0    0        0      181       3  878962392  
8        0         0    0        0       95       5  879197329  
9        0         0    0        0      268       3  875742341  
10       0         0    0        0      189       5  893264174  
11       0         0    0        0      145       3  882181396  
12       0         0    0        0      158      

In [31]:
merged_df.tail()

movie id                          movie title release date  Action  \
99985      1678                          Mat' i syn      6-Feb-98       0   
99986      1679                           B. Monkey      6-Feb-98       0   
99987      1680                       Sliding Doors      1-Jan-98       0   
99988      1681                        You So Crazy      1-Jan-94       0   
99989      1682  Scream of Stone (Schrei aus Stein)      8-Mar-96       0   

       Adventure  Animation  Childrens  Comedy  Crime  Documentary  ...  \
99985          0          0          0       0      0            0  ...   
99986          0          0          0       0      0            0  ...   
99987          0          0          0       0      0            0  ...   
99988          0          0          0       1      0            0  ...   
99989          0          0          0       0      0            0  ...   

       Musical  Mystery  Romance  Sci-Fi  Thriller  War  Western  user id  \
99985        0        0        0       0         0    0        0      863   
99986        0        0        1       0         1    0        0      863   
99987        0        0        1       0         0    0        0      863   
99988        0        0        0       0         0    0        0      896   
99989        0        0        0       0         0    0        0      916   

       rating  timestamp  
99985       1  889289570  
99986       3  889289491  
99987       2  889289570  
99988       3  887160722  
99989       3  880845755  

[5 rows x 24 columns]

In [32]:
merged_df.shape

(99990, 24)

In [33]:
# 2. groupby() the "movie title", and "rating"
# and name it rating_count
rating_count = merged_df.groupby(['movie title'])['rating'].count().reset_index()
rating_count.head(10)

movie title  rating
0                    'Til There Was You        9
1                                 1-900        5
2                        101 Dalmatians      109
3                          12 Angry Men      125
4                                    187      41
5                  2 Days in the Valley       93
6          20,000 Leagues Under the Sea       72
7                 2001: A Space Odyssey      259
8  3 Ninjas: High Noon At Mega Mountain        5
9                         39 Steps, The       59

In [34]:
# 3. get the average of all the ratings for each movie
# and name it avg_rating
avg_rating = merged_df.groupby(['movie title'])['rating'].mean().sort_values(ascending=False).reset_index()
avg_rating.head()

movie title  rating
0                      Great Day in Harlem, A      5.0
1                                 Prefontaine      5.0
2                      Someone Else's America      5.0
3  Entertaining Angels: The Dorothy Day Story      5.0
4         Marlene Dietrich: Shadow and Light (     5.0

In [35]:
# 4. filter the movies with more than 100 reviews
# and name it movie_100_reviews
movie_100_reviews = rating_count[rating_count['rating']>100]['movie title']
movie_100_reviews.head()

2            101 Dalmatians 
3              12 Angry Men 
7     2001: A Space Odyssey 
15           Absolute Power 
16               Abyss, The 
Name: movie title, dtype: object

In [36]:
avg_rating[avg_rating['movie title'].isin(movie_100_reviews)]

movie title    rating
15               Close Shave, A   4.491071
16             Schindler's List   4.466443
17          Wrong Trousers, The   4.466102
18                   Casablanca   4.456790
20    Shawshank Redemption, The   4.445230
...                          ...       ...
1261                      Spawn   2.615385
1277              Event Horizon   2.574803
1291                      Crash   2.546875
1357              Jungle2Jungle   2.439394
1392             Cable Guy, The   2.339623

[334 rows x 2 columns]

In [37]:
top_25_movies = avg_rating[avg_rating['movie title'].isin(movie_100_reviews)].head(25)
top_25_movies

movie title    rating
15                                    Close Shave, A   4.491071
16                                  Schindler's List   4.466443
17                               Wrong Trousers, The   4.466102
18                                        Casablanca   4.456790
20                         Shawshank Redemption, The   4.445230
21                                       Rear Window   4.387560
22                               Usual Suspects, The   4.385768
23                                         Star Wars   4.358491
24                                      12 Angry Men   4.344000
28                                      Citizen Kane   4.292929
30                             To Kill a Mockingbird   4.292237
31                   One Flew Over the Cuckoo's Nest   4.291667
32                         Silence of the Lambs, The   4.289744
33                                North by Northwest   4.284916
34                                    Godfather, The   4.283293
35                                    Secrets & Lies   4.265432
36                                 Good Will Hunting   4.262626
37                         Manchurian Candidate, The   4.259542
38  Dr. Strangelove or: How I Learned to Stop Worr...  4.252577
39                           Raiders of the Lost Ark   4.252381
40                                           Vertigo   4.251397
44                                           Titanic   4.245714
45                                Lawrence of Arabia   4.231214
47                               Maltese Falcon, The   4.210145
48                          Empire Strikes Back, The   4.204360

In [38]:
len(top_25_movies)

25

### What is the relationship between movie genres and gender?

In [39]:
all_data = merged_df.merge(user, on='user id')

In [40]:
genre_by_gender = all_data.groupby('gender').sum().loc[:,'Action':'Western']

In [41]:
genre_by_gender['total'] = all_data['gender'].value_counts()

In [42]:
genre_by_gender.T

gender           F      M
Action        5442  20147
Adventure     3141  10612
Animation      995   2610
Childrens     2232   4950
Comedy        8068  21764
Crime         1794   6261
Documentary    187    571
Drama        11008  28887
Fantasy        363    989
Film-Noir      385   1348
Horror        1197   4120
Musical       1442   3512
Mystery       1314   3931
Romance       5858  13603
Sci-Fi        2629  10101
Thriller      5086  16786
War           2189   7209
Western        371   1483
total        25738  74252

In [43]:
(genre_by_gender.div(genre_by_gender.total, axis=0)* 100).T

gender                F           M
Action        21.143834   27.133276
Adventure     12.203745   14.291871
Animation      3.865879    3.515057
Childrens      8.672002    6.666487
Comedy        31.346647   29.310995
Crime          6.970239    8.432096
Documentary    0.726552    0.769003
Drama         42.769446   38.904003
Fantasy        1.410366    1.331951
Film-Noir      1.495843    1.815439
Horror         4.650711    5.548672
Musical        5.602611    4.729839
Mystery        5.105292    5.294133
Romance       22.760121   18.320045
Sci-Fi        10.214469   13.603674
Thriller      19.760665   22.606798
War            8.504934    9.708829
Western        1.441448    1.997253
total        100.000000  100.000000

## Conclusion

List 15 insights here:
1. Adding T to a dataframe will transpose it, meaning that the rows and columns will be swapped. This is helpful to make a dataframe easier to look at if the dataframe happens to have more rows than columns or more columns than rows.
2. The sum() returns a sum, if used in a dataframe, it returns the sum values of each variable(columns).
3. Using the sort_values() method will sort the dataframe, this is set default to sort ascending.
4. For sort_values() method to sort descending, it has to have ascending set to False. Which would be: ascending=False. 
5. The loc() is used to index and search for certain values in the dataframe's variables. This is helpful for looking for specific values in the dataframe.
6. Setting axis to 1 will guide a function to just the columns of a dataframe. On the other hand, setting axis to 0 will guide a function to just the rows of a dataframe. By default, axis is always set to 0.
7. The merge() method will merge datasets together using a common variable. This is helpful when dealing with multiple sheets of datasets. 
8. The groupby() method will group variables in a dataframe. This can be done to create a new, smaller dataframe from the original dataframe. This is helpful for merged dataframes since merged dataframes tend to become extremely large after being merged. 
9. Using the value_counts() method will show the total value count of a variable in the dataframe.
10. Using isin() method will check if the dataframe has certain values. 
11. Using the reset_index() method will reset the index of a dataframe and give it a new, default index. 
12. Using mean() will return the average of the values.
13. When using head(), the amount of rows can be specified by adding a value into head(), like head(10). This would return the first 10 rows of the dataframe rather than 5.
14. According to the data mining on the relationship of gender and movie genres, males significantly review more movies than females do.
15. According to the data mining on the relationship of gender and movie genres (Percentage). The majority of females review drama movies, and the majority of males also review drama movies.